proposed/simplecache.md

```diff
@@ -0,0 +1,319 @@
+    const TTL_HOUR = 3600;
+    const TTL_DAY = 86400;
+
+    /**
+     * Fetched a value from the cache.
+     *
+     * @param string $key The unique key of this item in the cache
+     *
+     * @return mixed The value of the item from the cache, or null in case of cache miss
+     */
+    public function get($key);
+
+    /**
+     * Persist data in the cache, uniquely referenced by a key with an optional expiration TTL time.
+     *
+     * @param string $key The key of the item to store
+     * @param mixed $value The value of the item to store
+     * @param null|integer|DateInterval $ttl Optional. The TTL value of this item. If no value is sent and the driver supports TTL
+     *                                       then the library may set a default value for it or let the driver take care of that.
+     *
+     * @return bool True on success and false on failure
+     */
+    public function set($key, $value, $ttl = null);
+
+    /**
+     * Delete an item from the cache by its unique key
+     *
+     * @param string $key The unique cache key of the item to delete
+     *
+     * @return bool True on success and false on failure
+     */
+    public function delete($key);
+
+    /**
+     * Wipe clean the entire cache's keys
+     *
+     * @return bool True on success and false on failure
+     */
+    public function clear();
+
+    /**
+     * Obtain multiple cache items by their unique keys
+     *
+     * @param array|Traversable $keys A list of keys that can obtained in a single operation.
+     *
+     * @return array An array of key => value pairs. Cache keys that do not exist or are stale will have a value of null.
+     */
+    public function getMultiple($keys);
+
+    /**
+     * Persisting a set of key => value pairs in the cache, with an optional TTL.
+     *
+     * @param array|Traversable         $items An array of key => value pairs for a multiple-set operation.
+     * @param null|integer|DateInterval $ttl   Optional. The amount of seconds from the current time that the item will exist in the cache for.
+     *                                         If this is null then the cache backend will fall back to its own default behaviour.
+     *
+     * @return bool True on success and false on failure
+     */
+    public function setMultiple($items, $ttl = null);
+
+    /**
+     * Delete multiple cache items in a single operation
+     *
+     * @param array|Traversable $keys The array of string-based keys to be deleted
+     *
+     * @return bool True on success and false on failure
+     */
+    public function deleteMultiple($keys);
+
+    /**
+     * Increment a value atomically in the cache by its step value, which defaults to 1
+     *
+     * @param string  $key  The cache item key
+     * @param integer $step The value to increment by, defaulting to 1
+     *
+     * @return int|bool The new value on success and false on failure
+     */
+    public function increment($key, $step = 1);
```

Comment:
> @Ocramius yes it needs to be atomic otherwise how do you trust a rate limiter or any counter under high load?

Well, I wouldn't expect you to do them in PHP for sure :-P
Also, counting 100 or 102 is not a big problem with rate limiters, unless we are in a money-sensitive context (and then you need transactions anyway).

 > why bother with a SQL transaction?

When rate limiting needs to be precise, then there is usually a strong reasoning in order for it to be precise. In scenarios where consistency is critical, relying on a cache seems to be a mistake, from my PoV.

With typical SQL RDBMSs, It can be done with a single statement (`UPSERT`, usually, but of course support for it isn't widespread), with implicit transactions.

Anyway, this is going to be implemented in a likely non-atomic way in most adapters, implementation being what I pasted above, unless there are other 50000 lines around it handling concurrency:

```php
$cache->set('counter', ((int) $cache->get('counter')) + 1)
```

 > It's just different backends for different purposes IMO.

Right, and that's also my point here: counters are not the purpose of a cache. You can use a cache for counters by accepting its big limitations, or use a specific implementation (redis, in this case), if it needs to be an atomic operation.

I know that you use this stuff for packagist, but is caching a precise number really so important that you'd want an atomic implementation as part of the interface? What kind of actual damage could dogpiling and race conditions cause in this scenario? And again, assuming the atomicity requirement is dropped, then why not just make it via the one-liner I just wrote (non atomic)?